bundle/libraries/upload.go

```diff
@@ -0,0 +1,268 @@
+					if !IsLocalLibrary(&lib) {
+						continue
+					}
+
+					p := dyn.NewPath(
+						dyn.Key("resources"),
+						dyn.Key("jobs"),
+						dyn.Key(i),
+						dyn.Key("tasks"),
+						dyn.Index(j),
+						dyn.Key("for_each_task"),
+						dyn.Key("task"),
+						dyn.Key("libraries"),
+						dyn.Index(l),
+					)
+
+					if lib.Whl != "" {
+						libs[lib.Whl] = append(libs[lib.Whl], configLocation{
+							configPath: p.Append(dyn.Key("whl")),
+							location:   b.Config.GetLocation(p.String()),
+						})
+					}
+
+					if lib.Jar != "" {
+						libs[lib.Jar] = append(libs[lib.Jar], configLocation{
+							configPath: p.Append(dyn.Key("jar")),
+							location:   b.Config.GetLocation(p.String()),
+						})
+					}
+				}
+			}
+		}
+
+		for j, env := range job.Environments {
+			if env.Spec == nil {
+				continue
+			}
+
+			for k, dep := range env.Spec.Dependencies {
+				if !IsLibraryLocal(dep) {
+					continue
+				}
+
+				p := dyn.NewPath(
+					dyn.Key("resources"),
+					dyn.Key("jobs"),
+					dyn.Key(i),
+					dyn.Key("environments"),
+					dyn.Index(j),
+					dyn.Key("spec"),
+					dyn.Key("dependencies"),
+					dyn.Index(k),
+				)
+
+				libs[dep] = append(libs[dep], configLocation{
+					configPath: p,
+					location:   b.Config.GetLocation(p.String()),
+				})
+			}
+		}
+	}
+
+	for key, artifact := range b.Config.Artifacts {
+		for i, file := range artifact.Files {
+			p := dyn.NewPath(
+				dyn.Key("artifacts"),
+				dyn.Key(key),
+				dyn.Key("files"),
+				dyn.Index(i),
+			)
+
+			libs[file.Source] = append(libs[file.Source], configLocation{
+				configPath: p.Append(dyn.Key("remote_path")),
+				location:   b.Config.GetLocation(p.String()),
+			})
+		}
+	}
+
+	return libs
+}
+
+func (u *upload) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagnostics {
+	uploadPath, err := GetUploadBasePath(b)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	// If the client is not initialized, initialize it
+	// We use client field in mutator to allow for mocking client in testing
+	if u.client == nil {
+		filer, err := GetFilerForLibraries(b.WorkspaceClient(), uploadPath)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		u.client = filer
+	}
+
+	var diags diag.Diagnostics
+
+	libs := collectLocalLibraries(b)
+	errs, errCtx := errgroup.WithContext(ctx)
+	for source := range libs {
+		errs.Go(func() error {
+			return UploadFile(errCtx, source, u.client)
+		})
+	}
+
+	if err := errs.Wait(); err != nil {
+		return diag.FromErr(err)
+	}
+
+	// Update all the config paths to point to the uploaded location
+	for source, locations := range libs {
+		err = b.Config.Mutate(func(v dyn.Value) (dyn.Value, error) {
+			remotePath := path.Join(uploadPath, filepath.Base(source))
+
+			// If the remote path does not start with /Workspace or /Volumes, prepend /Workspace
+			if !strings.HasPrefix(remotePath, "/Workspace") && !strings.HasPrefix(remotePath, "/Volumes") {
+				remotePath = "/Workspace" + remotePath
+			}
```

Comment:
I know we're already doing this but in theory this would break for the folks who still have the `/Workspace` workspace folder.